# Lista de ejercicios 2

1.Supongamos que $x$ es un vector numérico. **Explica en detalle**, como las siguientes expresiones son evaluadas y que valores toman:

* `sum(!is.na(x))`

Si `x` es un vector, `is.na(x)` es otro vector tal que su i-ésimo elemento es TRUE si el i-ésimo elemento de `x` es `NA`, en caso contrario es `FALSE`. `!is.na(x)` es un nuevo vector tal que su i-ésimo elemento es `TRUE` si el i-ésimo elemento de `is.na(x)` es `FALSE`, en caso contrario es `FALSE`. Al pasar el vector `!is.na(x)` como parámetro de `sum()`, esta función toma como `1` a `TRUE` y como `0` a `FALSE`. Luego, suma todos los `0s` y `1s`.

* `c(x,x[-(1:length(x))])`

`x[-(1:length(x))]` retorna un vector numérico de longitud `0` ya que indicamos con el "-" que no se tome ningún
elemento del vector `x`. Luego, concatenamos este vector con `x` y resulta el mismo `x`.

* `x[length(x) + 1]/length(x)`

`x[length(x) + 1]` retorna `NA` ya que el vector no tiene elementos más allá de su longitud. Luego, `NA` dividido con cualquier número resulta `NA`.

* `sum(x > mean(x))`

La expresión lógica `x > mean(x)` nos retorna un vector cuyo i-ésimo elemento es `TRUE` si es mayor que el promedio, en caso contrario es `FALSE`. Al pasar este vector como parámetro de `sum()` ocurre lo mismo que en la primera expresión.

2.La función seno hiperbólico es definida como
$$sinh x = \frac{e^x - e^{-x}}{2}$$
donde $e^x$ es la función exponencial. Usando solo operaciones aritméticas y la función `exp`, escribe una función `vectorizada` en R, que calcule el $sinh$.

In [1]:
sinhiper <- function(x) 
{
    (exp(x)-exp(-x))/2
}

3.Crea un vector de los valores de $e^x\cos(x)$ en $x = 3,3.1, 3.2, \dots, 6$.

In [2]:
x <- seq(3,6,by=0.1) #x = 3, 3.1, 3.2, … , 6
vector <- exp(x)*cos(x)
vector #mostramos los elementos del vector

[1] -19.884531 -22.178753 -24.490697 -26.773182 -28.969238 -31.011186
 [7] -32.819775 -34.303360 -35.357194 -35.862834 -35.687732 -34.685042
[13] -32.693695 -29.538816 -25.032529 -18.975233 -11.157417  -1.362099
[19]  10.632038  25.046705  42.099201  61.996630  84.929067 111.061586
[25] 140.525075 173.405776 209.733494 249.468441 292.486707 338.564378
[31] 387.360340

4.Usa la función `paste` para crear el siguiente vector de caracteres de longitud $30$:

* ("vector 1", "vector 2", $\dots$, "vector 30"). Notar el espacio entre `vector` y el número.

In [3]:
paste("vector",1:30)

[1] "vector 1"  "vector 2"  "vector 3"  "vector 4"  "vector 5"  "vector 6" 
 [7] "vector 7"  "vector 8"  "vector 9"  "vector 10" "vector 11" "vector 12"
[13] "vector 13" "vector 14" "vector 15" "vector 16" "vector 17" "vector 18"
[19] "vector 19" "vector 20" "vector 21" "vector 22" "vector 23" "vector 24"
[25] "vector 25" "vector 26" "vector 27" "vector 28" "vector 29" "vector 30"

* ("fn1", "fn2", \dots, "fn30"). En este caso no hay espacio entre `fn` y el número.

In [4]:
paste("fn",1:30,sep = "")

[1] "fn1"  "fn2"  "fn3"  "fn4"  "fn5"  "fn6"  "fn7"  "fn8"  "fn9"  "fn10"
[11] "fn11" "fn12" "fn13" "fn14" "fn15" "fn16" "fn17" "fn18" "fn19" "fn20"
[21] "fn21" "fn22" "fn23" "fn24" "fn25" "fn26" "fn27" "fn28" "fn29" "fn30"

5.¿Qué retorna `DF <- D(expression(cos(x)/sin(x)), "x")`?

In [5]:
#Retorna la derivada de la expresión respecto de x
DF <- D(expression(cos(x)/sin(x)),"x")
DF

-(sin(x)/sin(x) + cos(x) * cos(x)/sin(x)^2)

Si ejecutamos el comando `x <- pi/4`, ¿que retorna `eval(DF)`?

In [6]:
x <- pi/4

eval(DF)  #evalua DF para el valor establecido previamente para x

[1] -2

6.¿Puedes explicar los dos siguientes resultados?

`x <- c(0,7,8)`

* `x[0.9999999999999999]` 

Retorna `numeric(0)`. R detecta al índice `0.9999999999999999` (con 16 9s) menor que `1`. Así, `x[0.9999999999999999]` se entiende como `x[0]` por lo cual nos retorna `numeric(0)`.

* `x[0.99999999999999999]`

Retorna `0`. R detecta al índice `0.99999999999999999` (con 17 9s) como `1`. Así, `x[0.99999999999999999]` se entiende como `x[1]` por lo cual nos retorna `0` (el primer elemento de x).

7.Usando `rep()` y `seq()`, crea los siguientes vectores:

* `0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4`

In [7]:
rep(0:4,each=5)

[1] 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4

* `1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5`

In [8]:
rep(1:5,5)

[1] 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5

8.Usando la función `cumprod` u otra relacionada, calcula:
$$1 + \frac{2}{3}+ \biggl(\frac{2}{3}\frac{4}{5} \biggr) + \biggl(\frac{2}{3}\frac{4}{5}\frac{6}{7} \biggr) + \cdots + \biggl(\frac{2}{3}\frac{4}{5}\cdots \frac{38}{39} \biggr)$$

In [9]:
#cumprod retorna un vector cuyos elementos son el 
#producto acumulado de los elementos del argumento
sum(cumprod(seq(2,38,by=2))/cumprod(seq(3,39,by=2))) + 1

[1] 6.976346

9.En primer lugar, lanzamos una moneda con dos posibles resultados: o bien cae cara o sello. A continuación, tiramos un dado con $6$ posibles resultados: $1,2,3,4,5$ y $6$. Genera todos los resultados posibles de esas acciones.

In [10]:
expand.grid(moneda = c("C","S"),dado=c(1:6))

moneda dado
1  C      1   
2  S      1   
3  C      2   
4  S      2   
5  C      3   
6  S      3   
7  C      4   
8  S      4   
9  C      5   
10 S      5   
11 C      6   
12 S      6

10.Tienes un conjunto de colores para elegir:

```
colores <- c("rojo", "azul", "verde", "blanco", "negro", "amarillo") 
```

Ahora debes  elegir $3$ colores y no puedes elegir el mismo color más de una vez. Enumera todas las combinaciones posibles.

In [11]:
colores <- c("rojo", "azul", "verde", "blanco", "negro", "amarillo")

#Elegir 3 colores sin repetir
combn(colores,3)

[,1]  [,2]   [,3]  [,4]     [,5]   [,6]  [,7]     [,8]   [,9]     [,10]   
[1,] rojo  rojo   rojo  rojo     rojo   rojo  rojo     rojo   rojo     rojo    
[2,] azul  azul   azul  azul     verde  verde verde    blanco blanco   negro   
[3,] verde blanco negro amarillo blanco negro amarillo negro  amarillo amarillo
     [,11]  [,12] [,13]    [,14]  [,15]    [,16]    [,17]  [,18]    [,19]   
[1,] azul   azul  azul     azul   azul     azul     verde  verde    verde   
[2,] verde  verde verde    blanco blanco   negro    blanco blanco   negro   
[3,] blanco negro amarillo negro  amarillo amarillo negro  amarillo amarillo
     [,20]   
[1,] blanco  
[2,] negro   
[3,] amarillo

11.Tienes el mismo conjunto de colores y puedes elegir $1$, $2$ o $3$. Haz una lista de todas las opciones posibles.

In [12]:
#Elegir 1 color sin repetir
combn(colores,1)

[,1] [,2] [,3]  [,4]   [,5]  [,6]    
[1,] rojo azul verde blanco negro amarillo

In [13]:
#Elegir 2 colores sin repetir
combn(colores,2)

[,1] [,2]  [,3]   [,4]  [,5]     [,6]  [,7]   [,8]  [,9]     [,10]  [,11]
[1,] rojo rojo  rojo   rojo  rojo     azul  azul   azul  azul     verde  verde
[2,] azul verde blanco negro amarillo verde blanco negro amarillo blanco negro
     [,12]    [,13]  [,14]    [,15]   
[1,] verde    blanco blanco   negro   
[2,] amarillo negro  amarillo amarillo

In [14]:
#Elegir 3 colores sin repetir
combn(colores,3)

[,1]  [,2]   [,3]  [,4]     [,5]   [,6]  [,7]     [,8]   [,9]     [,10]   
[1,] rojo  rojo   rojo  rojo     rojo   rojo  rojo     rojo   rojo     rojo    
[2,] azul  azul   azul  azul     verde  verde verde    blanco blanco   negro   
[3,] verde blanco negro amarillo blanco negro amarillo negro  amarillo amarillo
     [,11]  [,12] [,13]    [,14]  [,15]    [,16]    [,17]  [,18]    [,19]   
[1,] azul   azul  azul     azul   azul     azul     verde  verde    verde   
[2,] verde  verde verde    blanco blanco   negro    blanco blanco   negro   
[3,] blanco negro amarillo negro  amarillo amarillo negro  amarillo amarillo
     [,20]   
[1,] blanco  
[2,] negro   
[3,] amarillo

12.Responde lo siguiente:

(a) Sea $X$ el número de `"unos"` obtenidos en 12 lanzamientos  de un dado. Entonces $X$ tiene una  distribución Binomial $(n=12, p=1/3)$ . Calcule una tabla de probabilidades binomiales para $x=0, 1, \dots ,12$ por dos métodos:

* Usando la fórmula para la densidad:
$$P(X = K)=\binom{n}{k}p^{k}(1 -p)^{n -k}$$
y aritmética en R. Usa `0:12` para la secuencia de $x$ valores y la función `choose` para calcular los coeficientes binomiales $\binom{n}{k}$.

In [15]:
res1 <- choose(n=12,k=0:12)*((1/3)**(0:12))*((2/3)**(12:0))
res1

[1] 7.707347e-03 4.624408e-02 1.271712e-01 2.119520e-01 2.384460e-01
 [6] 1.907568e-01 1.112748e-01 4.768921e-02 1.490288e-02 3.311751e-03
[11] 4.967626e-04 4.516023e-05 1.881676e-06

* Usando la función `dbinom` de R.

In [16]:
res2 <- dbinom(x=0:12,size=12,prob=1/3)
res2

[1] 7.707347e-03 4.624408e-02 1.271712e-01 2.119520e-01 2.384460e-01
 [6] 1.907568e-01 1.112748e-01 4.768921e-02 1.490288e-02 3.311751e-03
[11] 4.967626e-04 4.516023e-05 1.881676e-06

Compara los resultados con ambos métodos:

In [17]:
#Para comparar usamos all.equal()
all.equal(res1,res2)

[1] TRUE

(b) Sea $X$ el número de `"unos"` obtenidos en 12 lanzamientos  de un dado. Entonces $X$ tiene una  distribución Binomial $(n=12, p=1/3)$. Calcule el CDF para $x=0, 1, \dots ,12$ por dos métodos:


* Usando la función `cumsum` y el resultado del ejercicio anterior.

In [18]:
cumsum(res2)

[1] 0.007707347 0.053951426 0.181122646 0.393074678 0.631520714 0.822277544
 [7] 0.933552360 0.981241568 0.996144445 0.999456196 0.999952958 0.999998118
[13] 1.000000000

* Con el uso de la función `pbinom`.

In [19]:
pbinom(q=0:12,size=12,prob=1/3)

[1] 0.007707347 0.053951426 0.181122646 0.393074678 0.631520714 0.822277544
 [7] 0.933552360 0.981241568 0.996144445 0.999456196 0.999952958 0.999998118
[13] 1.000000000

¿Qué es $P(X >7)$?

In [20]:
#P(X>7) = 1 - P(X<=7)
1-sum(dbinom(x=0:7,size=12,prob=1/3))

[1] 0.01875843

13.Responde lo siguiente:

(a) ¿Cuál es la probabilidad de que una persona sea menor o igual a 1.90 m? Usa `pnorm`.

In [21]:
#P[X<=1.9]
pnorm(1.90)

[1] 0.9712834

(b) ¿Cuál es la probabilidad de que una persona sea más alta o igual a 1.60 m?. Usa `pnorm`.

In [22]:
#P[X>=1.6]
1 - pnorm(1.60)

[1] 0.05479929

14.Cuál es la probabilidad de que una persona espere menos de $10$ minutos?. Utiliza `pexp`.

In [23]:
pexp(10)

[1] 0.9999546

15.Ejecuta las siguientes líneas que crean dos vectores de enteros aleatorios que se eligen con reemplazo de los enteros $0, 1,\dots, 999$ . Ambos vectores tienen una longitud de $250$.

In [24]:
set.seed(50) 
#dada la misma semilla (en este caso 50), se obtiene la misma secuencia
#tanto de xVec como de yVec
xVec <- sample(0:999, 250, replace=T)
yVec <- sample(0:999, 250, replace=T)